integrators/integrators_body.hpp

```diff
@@ -14,80 +14,254 @@
 #include "integrators/symmetric_linear_multistep_integrator.hpp"
 #include "integrators/symplectic_runge_kutta_nyström_integrator.hpp"
 
-#define PRINCIPIA_CASE_SLMS(kind, method)                      \
-  case serialization::FixedStepSizeIntegrator::kind:           \
-    return SymmetricLinearMultistepIntegrator<methods::method, \
-                                              typename ODE::Position>()
+#define PRINCIPIA_CASE_SLMS(kind, method, action)      \
+  case serialization::FixedStepSizeIntegrator::kind: { \
+    action(method);                                    \
+  }
 
-// We do not deserialize an SPRK per se, but only when it is converted to an
-// SRKN.  The reason is that an SPRK is for a different kind of equation than
-// an SRKN, so the two would return different types.  If we ever need to do
-// this we will need to specialize ReadFromMessage somehow.
-#define PRINCIPIA_CASE_SPRK(kind, method)                                 \
-  case serialization::FixedStepSizeIntegrator::kind:                      \
-    do {                                                                  \
-      CHECK(message.has_composition_method());                            \
-      return SprkAsSrknDeserializer<                                      \
-          ODE,                                                            \
-          methods::method,                                                \
-          methods::method::first_same_as_last>::ReadFromMessage(message); \
-    } while (false)
-
-#define PRINCIPIA_CASE_SRKN(kind, method)                         \
-  case serialization::FixedStepSizeIntegrator::kind:              \
-    return SymplecticRungeKuttaNyströmIntegrator<methods::method, \
-                                                 typename ODE::Position>()
+#define PRINCIPIA_CASE_SPRK(kind, method, action)      \
+  case serialization::FixedStepSizeIntegrator::kind: { \
+    action(method);                                    \
+  }
+
+#define PRINCIPIA_CASE_SRKN(kind, method, action)      \
+  case serialization::FixedStepSizeIntegrator::kind: { \
+    action(method);                                    \
+  }
```

Comment:
*You are in a maze of twisty little macros, all alike.*

Perhaps these should be one macro.